pilot/pkg/model/authorization.go

```diff
@@ -66,42 +66,61 @@ type AuthorizationPoliciesResult struct {
 }
 
 // ListAuthorizationPolicies returns authorization policies applied to the workload in the given namespace.
-func (policy *AuthorizationPolicies) ListAuthorizationPolicies(namespace string, workload labels.Instance) AuthorizationPoliciesResult {
-	ret := AuthorizationPoliciesResult{}
+func (policy *AuthorizationPolicies) ListAuthorizationPolicies(selectionOpts WorkloadSelectionOpts) AuthorizationPoliciesResult {
+	configs := AuthorizationPoliciesResult{}
 	if policy == nil {
-		return ret
+		return configs
 	}
+	rootNamespace := policy.RootNamespace
+	namespace := selectionOpts.Namespace
+	workloadLabels := selectionOpts.WorkloadLabels
+	var lookupInNamespaces []string
 
-	var namespaces []string
-	if policy.RootNamespace != "" {
-		namespaces = append(namespaces, policy.RootNamespace)
-	}
-	// To prevent duplicate policies in case root namespace equals proxy's namespace.
-	if namespace != policy.RootNamespace {
-		namespaces = append(namespaces, namespace)
+	if namespace != rootNamespace {
+		// Only check the root namespace if the (workload) namespace is not already the root namespace
+		// to avoid double inclusion.
+		lookupInNamespaces = []string{rootNamespace, namespace}
+	} else {
+		lookupInNamespaces = []string{namespace}
 	}
 
-	for _, ns := range namespaces {
+	for _, ns := range lookupInNamespaces {
 		for _, config := range policy.NamespaceToPolicies[ns] {
 			spec := config.Spec
-			selector := labels.Instance(spec.GetSelector().GetMatchLabels())
-			if selector.SubsetOf(workload) {
-				switch config.Spec.GetAction() {
-				case authpb.AuthorizationPolicy_ALLOW:
-					ret.Allow = append(ret.Allow, config)
-				case authpb.AuthorizationPolicy_DENY:
-					ret.Deny = append(ret.Deny, config)
-				case authpb.AuthorizationPolicy_AUDIT:
-					ret.Audit = append(ret.Audit, config)
-				case authpb.AuthorizationPolicy_CUSTOM:
-					ret.Custom = append(ret.Custom, config)
-				default:
-					log.Errorf("ignored authorization policy %s.%s with unsupported action: %s",
-						config.Namespace, config.Name, config.Spec.GetAction())
+			var selector labels.Instance // NOTE: nil/empty selector matches all workloads
+			switch getPolicyMatcher(gvk.AuthorizationPolicy, config.Name, selectionOpts, spec) {
+			case policyMatchSelector:
+				selector = spec.GetSelector().GetMatchLabels()
+				if selector.SubsetOf(workloadLabels) {
+					configs = updateAuthorizationPoliciesResult(configs, config)
 				}
+			case policyMatchDirect:
+				configs = updateAuthorizationPoliciesResult(configs, config)
+				continue
```

Comment:
```suggestion
				continue
```
this does nothing (?)

Same with L101